src/Quizzer.API.BL/Facades/QuizFacade.cs

```diff
@@ -0,0 +1,167 @@
+            IEnumerable<QuizEntity> entities = repository.GetByPredicate(predicate);
+
+            IEnumerable<QuizListModel> models = entities.Select(p_mapper.Map<QuizListModel>);
+            return models;
+        }
+
+        public void StartQuiz(Guid userId, Guid quizId, DateTime time) {
+            using IUnitOfWork uow = p_unitOfWorkFactory.Create();
+
+            using IRepository<QuizEntity> quizRepository = uow.GetRepository<QuizEntity>();
+            using IRepository<UserEntity> userRepository = uow.GetRepository<UserEntity>();
+            using IRepository<CompletedQuizEntity> completedQuizRepository = uow.GetRepository<CompletedQuizEntity>();
+
+            QuizEntity quiz = quizRepository.GetById(quizId) ??
+                throw new ArgumentException("Quiz does not exist");
+
+            UserEntity user = userRepository.GetById(userId) ??
+                throw new ArgumentException("User does not exist");
+
+            CompletedQuizEntity? completedQuiz = completedQuizRepository.GetByPredicate(
+                q => q.UserId == userId && q.QuizId == quizId).SingleOrDefault();
+
+            if (completedQuiz is not null) throw new InvalidOperationException("Quiz has been started");
+
+            if (time < quiz.StartTime || quiz.EndTime < time) throw new InvalidOperationException("Quiz is not opened");
+
+            completedQuizRepository.Insert(new CompletedQuizEntity() {
+                Id = Guid.NewGuid(),
+                Score = 0,
+                InProgress = true
+            });
+
+            uow.Commit();
+        }
+
+        public void CompleteQuiz(Guid quizId, Guid userId) {
+            using IUnitOfWork uow = p_unitOfWorkFactory.Create();
+            using IRepository<CompletedQuizEntity> completedQuizRepository = uow.GetRepository<CompletedQuizEntity>();
+
+            CompletedQuizEntity? completedQuiz = completedQuizRepository.GetByPredicate(
+                q => q.QuizId == quizId && q.UserId == userId,
+                nameof(CompletedQuizEntity.Quiz), nameof(CompletedQuizEntity.User)).SingleOrDefault()
+                ?? throw new InvalidOperationException("Quiz has not been started");
+
+            if (!completedQuiz.InProgress) throw new InvalidOperationException("Quiz has been completed");
+
+            completedQuiz.InProgress = false;
+
+            using IRepository<UserAnswerEntity> userAnswerRepository = uow.GetRepository<UserAnswerEntity>();
+            IEnumerable<UserAnswerEntity> answers = userAnswerRepository.GetByPredicate(
+                a => a.UserId == userId && a.Question!.QuizId == quizId,
+                nameof(UserAnswerEntity.Question), $"{nameof(UserAnswerEntity.Question)}.{nameof(QuestionEntity.Quiz)}");
+
+            completedQuiz.Score = answers.Select(a => a.AnswerScore).Sum();
+            completedQuizRepository.Update(completedQuiz);
+
+            uow.Commit();
+        }
+
+        public void AddQuestion(Guid questionId, Guid quizId, DateTime time) {
```

Comment:
I am not sure how will this work. Because `QuestionEntity` has reuqired `QuestionId` and you are getting question from database so which Id in the database will be stored in this `QuestionEntity`?

Reply:
I do not quite understand your complaint, In the `AddQuestion` method, I am just adding quiz Id for the question entity, thus setting the relation between the selected quiz and the selected question.

Reply:
Well i mean... You have a questionId... In this method you get the question by Id.. So it means that the qusetion is already in the database, right? Then the question needs to have in database collumn some quizId because it is required. But which Id it will be? Or am i wrong?

Reply:
Well yes, I mean that previous implementation when the Quiz Id was present inside model was maybe better, but I will try to edit this to fit the usecase